DynamicTablesPkg/Library/Acpi/X64/SsdtSerialPortFixupLib/SsdtSerialPortFixupLib.c

```diff
@@ -0,0 +1,518 @@
+      DEBUG ((
+        DEBUG_ERROR,
+        "ERROR: SSDT-SERIAL-PORT-FIXUP: Failed to generate UartSerialBus RD node."
+        " Status = %r\n",
+        Status
+        ));
+      return Status;
+    }
+  }
+
+  if (SerialPortInfo->PortSubtype == EFI_ACPI_DBG2_PORT_SUBTYPE_SERIAL_16550_WITH_GAS) {
+    Status = AmlCodeGenRdMemory32Fixed (
+               TRUE,
+               SerialPortInfo->BaseAddress & MAX_UINT32,
+               MIN_UART_ADDRESS_LENGTH,
+               CrsNode,
+               NULL
+               );
+    if (EFI_ERROR (Status)) {
+      DEBUG ((
+        DEBUG_ERROR,
+        "ERROR: SSDT-SERIAL-PORT-FIXUP: Failed to generate MMIO RD node."
+        " Status = %r\n",
+        Status
+        ));
+      return Status;
+    }
+
+    Status = AmlCodeGenRdIrq (
+               TRUE,
+               TRUE,
+               TRUE,
+               IrqList,
+               1,
+               CrsNode,
+               NULL
+               );
+    if (EFI_ERROR (Status)) {
+      DEBUG ((
+        DEBUG_ERROR,
+        "ERROR: SSDT-SERIAL-PORT-FIXUP: Failed to generate IRQ RD node."
+        " Status = %r\n",
+        Status
+        ));
+      return Status;
+    }
+  }
+
+  return EFI_SUCCESS;
+}
+
+/** Build a SSDT table describing the input serial port.
+
+  The table created by this function must be freed by FreeSsdtSerialTable.
+
+  @param [in]  AcpiTableInfo    Pointer to the ACPI table information.
+  @param [in]  SerialPortInfo   Serial port to describe in the SSDT table.
+  @param [in]  Name             The Name to give to the Device.
+                                Must be a NULL-terminated ASL NameString
+                                e.g.: "DEV0", "DV15.DEV0", etc.
+  @param [in]  Uid              UID for the Serial Port.
+  @param [out] Table            If success, pointer to the created SSDT table.
+
+  @retval EFI_SUCCESS            Table generated successfully.
+  @retval EFI_INVALID_PARAMETER  A parameter is invalid.
+  @retval EFI_NOT_FOUND          Could not find information.
+  @retval EFI_OUT_OF_RESOURCES   Could not allocate memory.
+**/
+EFI_STATUS
+EFIAPI
+BuildSsdtSerialPortTable (
+  IN  CONST CM_STD_OBJ_ACPI_TABLE_INFO       *AcpiTableInfo,
+  IN  CONST CM_ARCH_COMMON_SERIAL_PORT_INFO  *SerialPortInfo,
+  IN  CONST CHAR8                            *Name,
+  IN  CONST UINT64                           Uid,
+  OUT       EFI_ACPI_DESCRIPTION_HEADER      **Table
+  )
+{
+  AML_OBJECT_NODE_HANDLE  DeviceNode;
+  AML_OBJECT_NODE_HANDLE  ScopeNode;
+  AML_ROOT_NODE_HANDLE    RootNode;
+  CONST CHAR8             *NonBsaHid;
+  EFI_STATUS              Status;
+  EFI_STATUS              Status1;
+
+  ASSERT (AcpiTableInfo != NULL);
+  ASSERT (SerialPortInfo != NULL);
+  ASSERT (Name != NULL);
+  ASSERT (Table != NULL);
+
+  // Validate the Serial Port Info.
+  Status = ValidateSerialPortInfo (SerialPortInfo, 1);
+  if (EFI_ERROR (Status)) {
+    return Status;
+  }
+
+  Status = AmlCodeGenDefinitionBlock (

```

Comment:
Should be `AddSsdtAcpiHeader()` I think:
- to manage the OemRevision
- in case this replaces the Common generator

Reply:
Ok right understood